scipy/optimize/__slsqp.h

```diff
@@ -0,0 +1,405 @@
+ *  the code currently archived at http://www.netlib.org/toms/733 under the
+ *  BSD license, the ACM hereby releases this code under the BSD license.
+ *
+ *  Regards,
+ *
+ *  Deborah Cotton, Copyright & Permissions
+ *  ACM Publications
+ *  2 Penn Plaza, Suite 701**
+ *  New York, NY 10121-0701
+ *  permissions@acm.org
+ *  212.869.7440 ext. 652
+ *  Fax. 212.869.0481
+ *  ------
+*/
+
+#ifndef __SLSQPLIB_H
+#define __SLSQPLIB_H
+
+#define PY_SSIZE_T_CLEAN
+#include "Python.h"
+#include "numpy/arrayobject.h"
+
+#define PYERR(errobj,message) {PyErr_SetString(errobj,message); return NULL;}
+static PyObject* slsqp_error;
+
+#include <math.h>
+#include "__nnls.h"
+
+void daxpy_(int* n, double* sa, double* sx, int* incx, double* sy, int* incy);
+double ddot_(int* n, double* dx, int* incx, double* dy, int* incy);
+void dgelsy_(int* m, int* n, int* nrhs, double* a, int* lda, double* b, int* ldb, int* jpvt, double* rcond, int* rank, double* work, int* lwork, int* info);
+void dgemv_(char* trans, int* m, int* n, double* alpha, double* a, int* lda, double* x, int* incx, double* beta, double* y, int* incy);
+void dgeqr2_(int* m, int* n, double* a, int* lda, double* tau, double* work, int* info);
+void dgeqrf_(int* m, int* n, double* a, int* lda, double* tau, double* work, double* lwork, int* info);
+void dgerq2_(int* m, int* n, double* a, int* lda, double* tau, double* work, int* info);
+void dlarf_(char* side, int* m, int* n, double* v, int* incv, double* tau, double* c, int* ldc, double* work);
+void dlarfgp_(int* n, double* alpha, double* x, int* incx, double* tau);
+void dlartgp_(double* f, double* g, double* cs, double* sn, double* r);
+double dnrm2_(int* n, double* x, int* incx);
+void dorm2r_(char* side, char* trans, int* m, int* n, int* k, double* a, int* lda, double* tau, double* c, int* ldc, double* work, int* info);
+void dormr2_(char* side, char* trans, int* m, int* n, int* k, double* a, int* lda, double* tau, double* c, int* ldc, double* work, int* info);
+void dscal_(int* n, double* da, double* dx, int* incx);
+void dtpmv_(char* uplo, char* trans, char* diag, int* n, double* ap, double* x, int* incx);
+void dtpsv_(char* uplo, char* trans, char* diag, int* n, double* ap, double* x, int* incx);
+void dtrsm_(char* side, char* uplo, char* transa, char* diag, int* m, int* n, double* alpha, double* a, int* lda, double* b, int* ldb);
+void dtrsv_(char* uplo, char* trans, char* diag, int* n, double* a, int* lda, double* x, int* incx);
+
+struct SLSQP_static_vars {
+    double acc, alpha, f0, gs, h1, h2, h3, h4, t, t0, tol;
+    int exact, inconsistent, reset, iter, itermax, line, m, meq, mode, n;
+};
+
+void __slsqp_body(struct SLSQP_static_vars* S, double* funx, double* gradx, double* C, double* d, double* sol, double* mult, double* xl, double* xu, double* buffer, int* indices);
+
+
+static PyObject*
+nnls(PyObject* Py_UNUSED(dummy), PyObject* args) {
+
+    int maxiter, info = 0;
```

Comment:
Should implementations be moved to `__slsqp.c`? That's where I would expect them, and this header is only included in that one C file. `_nnls.h` looks more pleasant to read.

Reply:
This is actually the input parsing that is coming from the user. So these values are populated by the `PyArg_ParseTuple` function. Then we call the actual low level function after a bit of sanity checks etc. Am I understanding your comment correctly?

Reply:
What I meant is that normally only signatures, includes, header guards go in header files, and the actual code goes in the source files. It's not always consistent in our code base, but if you grep for `PyInit_` for example, the vast majority is in `.c`/`.cc` files and not in the header. Input argument parsing is similar I think, usually in the sources not the header.

`__nnls.h` is quite minimal, and that seems more desirable in principle. 

Anyway, if you want to keep it like this, please don't let me stop you. It just seemed worth pointing out.

Reply:
Ah OK, I understand now. So the previous practice was to create a `<name>module.c` to handle all this and indeed the header was just to deal with the declarations. I could definitely have done that way, now that you put it like that. 

Hmm then I made all of them wrong so far. Let me get these in as is, then I'll make a clean up session for all of them.